New package: Task Herder

cwillisf opened the pull request:
Sorry for assigning this to so many people, but I'm not sure who has time & scope to review this. If you don't, no worries! If you do, come see the cool bike shed I made! :sweat_smile: 

### Resolves

In support of POD-238

### Proposed Changes

Extract `TaskQueue` from `scratch-vm` into its own package, `@scratch/task-herder`, with some improvements. The new package is still part of the `scratch-editor` mono-repository.

### Reason for Changes

I want to reuse this code in `scratch-storage`, which implies moving it to a common location.

### Test Coverage

There are 6 test suites totalling 100% coverage:
- Basics like constructing an instance and exceeding simple limits
- Task cancellation
- Rate limiting
- Concurrency limit=1 (serialized / non-concurrent tasks)
- Concurrency limit=2 (some concurrency)
- Rate limiting mixed with concurrency

Since this is the first `scratch-editor` component using these build tools, I also wanted to check the CI behavior. To do that, I intentionally checked in failing tests during the draft phase of this PR. Check the GHA history of this branch to see how that went. These tests prompted me to make some changes, but I'm happy with where it ended up.

cwillisf committed:
chore: add skeleton for task-herder library

cwillisf committed:
style: use eslint-config-scratch@12 recommended

cwillisf committed:
build: activate library mode

cwillisf committed:
feat!: first pass TokenBucket implementation

BREAKING CHANGE: the code in `scratch-vm/src/util/task-queue` as been
migrated to `@scratch/task-herder` and any references will need to be
updated. The new API supports all old use cases, but some terminology
has changed.

cwillisf committed:
refactor(deps): externalize dependencies

cwillisf committed:
refactor: rename limits for better self-documentation

cwillisf committed:
style: remove blanket indenting setting

cwillisf committed:
fix: wait for task completion so the concurrency limit works

cwillisf committed:
feat: clean up TaskHerder API and expose options and cancel reasons

cwillisf committed:
test: add tests for TaskHerder

cwillisf committed:
docs: add TaskHerder README.md

cwillisf committed:
fix: allow the AbortSignal to prevent a task from running

cwillisf committed:
test: add tests for canceling tasks

cwillisf committed:
ci: add path filter for task-herder

cwillisf committed:
ci: record junit output from Vitest

cwillisf committed:
test: include coverage

cwillisf committed:
test: add tests for task results and limit hacks

cwillisf committed:
chore: add @scratch/ scope to task-herder package

cwillisf committed:
fix: respond to automated review feedback

The `abortSignal` name is more descriptive, but using `signal` instead
means that the API is more consistent with `fetch`. Since I expect this
to be used with `fetch` quite a bit, I think the consistency is
worthwhile.

cwillisf committed:
build: generate `dist/*.d.ts` files during build

cwillisf committed:
refactor: split task record from queue

KManolov3 reviewed (approved):
After the new coat of paint this tool shed looks better than ever! 🙌

cwillisf committed:
refactor: use the cute name only for the libary, not the class

This also makes the class name less ambiguous: I'm working on an
additional feature that groups queues together, and I didn't want to end
up in a situation where I'm talking about a HerdHerd. That's the kind of
silliness that belongs in lunchtime conversations, not debugging
sessions :D
